### ISSUE TITLE
Crash in an actor with non-async implementation of an async protocol method

### ISSUE THREAD

**Issue by bealex** [swiftlang/swift#74126]
### Description

Here is the code, that I run as a command line utility in macOS (in "Reproduction")

There is an `async` in line 2, that is missing from the implementation. If it is present — I get a crash. If I remove it from there, crash disappears. Other things that are required for the crash to be reproduced:
 - `let bar: any Foo = Bar()` — here `any Foo` is required.
 - Implementation of `func testFunc(key: String) -> String?` should contain both reading and writing to the dictionary.

### Reproduction

```swift
protocol Foo: Actor {
    func testFunc(key: String) async -> String?
}

actor Bar: Foo {
    private var dictionary: [String: String] = [:]

    func testFunc(key: String) -> String? {
        dictionary[key] = "value for \(key)"
        return dictionary[key]
    }
}

let keys: [String] = [ "test1", "test2", "test3", "test4", "test5", "test6", "test7", ]
let bar: any Foo = Bar()

await withTaskGroup(of: Void.self) { group in
    for _ in 0 ... 10000 {
        group.addTask {
            guard let key = keys.randomElement() else { return }

            let value: String? = await bar.testFunc(key: key)
            print("got “\(value ?? "nil")”")
        }
    }
}
```

### Stack dump

```text
*** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[NSTaggedPointerString count]: unrecognized selector sent to instance 0x8000000000000000'
*** First throw call stack:
(
	0   CoreFoundation                      0x000000019b1af2ec __exceptionPreprocess + 176
	1   libobjc.A.dylib                     0x000000019ac96788 objc_exception_throw + 60
	2   CoreFoundation                      0x000000019b26156c -[NSObject(NSObject) __retain_OA] + 0
	3   CoreFoundation                      0x000000019b118f3c ___forwarding___ + 1580
	4   CoreFoundation                      0x000000019b118850 _CF_forwarding_prep_0 + 96
	5   libswiftCore.dylib                  0x00000001aadcff80 $sSD8_VariantV8setValue_6forKeyyq_n_xtF + 92
	6   libswiftCore.dylib                  0x00000001aad7ef70 $sSDyq_Sgxcis + 520
	7   TestConcurrencyAsync                0x0000000100002d54 $s20TestConcurrencyAsync3BarC8testFunc3keySSSgSS_tF + 416
	8   TestConcurrencyAsync                0x0000000100003024 $s20TestConcurrencyAsync3BarCAA3FooA2aDP8testFunc3keySSSgSS_tYaFTW + 36
	9   TestConcurrencyAsync                0x00000001000024e5 $s20TestConcurrencyAsyncyScGyytGzYaXEfU_yyYaYbcfU_TQ1_ + 1
	10  TestConcurrencyAsync                0x000000010000373d $sxIeghHr_xs5Error_pIegHrzo_s8SendableRzlTRTQ0_ + 1
	11  TestConcurrencyAsync                0x0000000100003899 $sxIeghHr_xs5Error_pIegHrzo_s8SendableRzlTRTATQ0_ + 1
	12  libswift_Concurrency.dylib          0x00000002631cf0f9 _ZL23completeTaskWithClosurePN5swift12AsyncContextEPNS_10SwiftErrorE + 1
)
got “value for test5”*** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[NSTaggedPointerString count]: unrecognized selector sent to instance 0x8000000000000000'
*** First throw call stack:
(
	0   CoreFoundation                      0x000000019b1af2ec __exceptionPreprocess + 176
	1   libobjc.A.dylib                     0x000000019ac96788 objc_exception_throw + 60
	2   CoreFoundation                      0x000000019b26156c -[NSObject(NSObject) __retain_OA] + 0
	3   CoreFoundation                      0x000000019b118f3c ___forwarding___ + 1580
	4   CoreFoundation                      0x000000019b118850 _CF_forwarding_prep_0 + 96
	5   libswiftCore.dylib                  0x00000001aadcff80 $sSD8_VariantV8setValue_6forKeyyq_n_xtF + 92
	6   libswiftCore.dylib                  0x00000001aad7ef70 $sSDyq_Sgxcis + 520
	7   TestConcurrencyAsync                0x0000000100002d54 $s20TestConcurrencyAsync3BarC8testFunc3keySSSgSS_tF + 416
	8   TestConcurrencyAsync                0x0000000100003024 $s20TestConcurrencyAsync3BarCAA3FooA2aDP8testFunc3keySSSgSS_tYaFTW + 36
	9   TestConcurrencyAsync                0x00000001000024e5 $s20TestConcurrencyAsyncyScGyytGzYaXEfU_yyYaYbcfU_TQ1_ + 1
	10  TestConcurrencyAsync                0x000000010000373d $sxIeghHr_xs5Error_pIegHrzo_s8SendableRzlTRTQ0_ + 1
	11  TestConcurrencyAsync                0x0000000100003899 $sxIeghHr_xs5Error_pIegHrzo_s8SendableRzlTRTATQ0_ + 1
	12  libswift_Concurrency.dylib          0x00000002631cf0f9 _ZL23completeTaskWithClosurePN5swift12AsyncContextEPNS_10SwiftErrorE + 1
)
```

### Expected behavior

No crash

### Environment

swift-driver version: 1.90.11.1 Apple Swift version 5.10 (swiftlang-5.10.0.13 clang-1500.3.9.4)
Target: arm64-apple-macosx14.0

**Comment by bealex** [comment]
Here is some more investigation from Swift forums user [vns](https://forums.swift.org/u/vns): https://forums.swift.org/t/an-actor-with-non-async-implementation-of-an-async-protocol-method-crashes-periodically/72247/17

**Comment by hooman** [comment]
It seems similar to this bug that @DougGregor fixed a couple of years ago:

https://github.com/apple/swift/issues/58517

**Comment by bealex** [comment]
> It seems similar to this bug that @DougGregor fixed a couple of years ago:
> 
> #58517

Wow. I tried to search for something similar in open issues and failed. Thanks! Looks very similar indeed.

**Comment by bealex** [comment]
Looks like it is fixed in the latest toolchain. It is still valid for the 5.10 though.

(from here: https://forums.swift.org/t/an-actor-with-non-async-implementation-of-an-async-protocol-method-crashes-periodically/72247/21)

> I think my statement about latest toolchain turned out to be a mistake. While I've been checking other things within it, I noticed that compilation command was wrong (it actually compiled against 5.10 version for some reason), so I just saw old SIL output. It is actually not crashing with latest toolchain, and has hop to Bar executor at the protocol witness level (in 5.10 this hop still missing, so there is an issue with that):

```
// protocol witness for Foo.testFunc(key:) in conformance Bar
sil private [transparent] [thunk] @$s3App3BarCAA3FooA2aDP8testFunc3keySSSgSS_tYaFTW : $@convention(witness_method: Foo) @async (@guaranteed String, @sil_isolated @guaranteed Bar) -> @owned Optional<String> {
// %0                                             // user: %4
// %1                                             // users: %4, %3, %2
bb0(%0 : $String, %1 : $Bar):
  hop_to_executor %1 : $Bar                       // id: %2
  %3 = class_method %1 : $Bar, #Bar.testFunc : (isolated Bar) -> (String) -> String?, $@convention(method) (@guaranteed String, @sil_isolated @guaranteed Bar) -> @owned Optional<String> // user: %4
  %4 = apply %3(%0, %1) : $@convention(method) (@guaranteed String, @sil_isolated @guaranteed Bar) -> @owned Optional<String> // user: %5
  return %4 : $Optional<String>                   // id: %5
} // end sil function '$s3App3BarCAA3FooA2aDP8testFunc3keySSSgSS_tYaFTW'
```

> And that's actually makes more sense, since actor has to guard executor by its own, so it was weird to me from the beginning that in other SIL hop was in other place, but never went for checking it.

**Comment by hborla** [comment]
@xedin fixed this in 6.0. Trying to find the exact PR...

**Comment by hborla** [comment]
Aha https://github.com/swiftlang/swift/pull/71480
